### PR TITLE
proto: Rename DomainType trait to Protobuf

### PR DESCRIPTION
**Pull request by thanethomson**
As discussed on our call yesterday, and towards fulfilling #654, in this PR I've:

1. Renamed the `DomainType` trait to `Protobuf` to clarify its purpose.
2. Kept it in the `tendermint-proto` crate, because it makes the most sense there.
3. Added a testable example of how to make use of the trait.

* [x] Referenced an issue explaining the need for the change
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG.md

### CODE REVIEW

File: tendermint/src/proposal.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@ use crate::{Error, Kind};
 use bytes::BufMut;
 use std::convert::{TryFrom, TryInto};
 use tendermint_proto::types::Proposal as RawProposal;
-use tendermint_proto::{DomainType, Error as DomainTypeError};
+use tendermint_proto::{Error as DomainTypeError, Protobuf};
```
Comment (greg-szabo) [inline tendermint/src/proposal.rs:20]:
```suggestion
use tendermint_proto::{Error as ProtobufError, Protobuf};
```
There are not many examples of this, but I guess it's worth renaming these too, for consistency.

**Review by greg-szabo — APPROVED** [review]
Looks good. Maybe we should also rename the cases where there are multiple Error crates imported and the one from tendermint_proto::Error is aliased as DomainTypeError. A quick search showed me this in vote.rs, proposal.rs, sign_proposal.rs and sign_vote.rs.

File: proto/src/lib.rs [hunk 2]
```diff
@@ -73,10 +73,174 @@ mod tendermint {
...
+///
+/// // We expect a validation error here
+/// assert!(MyDomainType::decode(invalid_raw_bytes.as_ref()).is_err());
+/// ```
+pub trait Protobuf<T: Message + From<Self> + Default>
+where
+    Self: Sized + Clone + TryFrom<T>,
+    <Self as TryFrom<T>>::Error: Into<BoxError>,
+{
+    /// Encodes the DomainType into a buffer.
```
Comment (adizere) [inline proto/src/lib.rs:168]: Nit:
Maybe we could stop referring to Self as "DomainType" in the comments for these methods. For example:

```suggestion
    /// Encodes the domain type (self) into a buffer.
```

Comment (thanethomson) [reply proto/src/lib.rs:168]: Yes you're right, I forgot to check the comments. Will update 👍

File: tendermint/src/proposal.rs (hunk 1 above)
Comment (thanethomson) [reply tendermint/src/proposal.rs]: Done 👍

**Review by greg-szabo — APPROVED** [review]
Good stuff.

I'm not sure if the word "equivalent" is the best to describe the similarities between the `Protobuf::encode()` and the `Prost::Message::encode()` functions. They have similar functionality that they implement differently. Maybe the word is part of some kind of documentation standard (you seem to have used the same template in multiple places) and I don't have a good enough vocabulary to give you a better alternative. Maybe "replacement"? `Protobuf::encode()` replaces the `Prost::Message::encode()` function with the added features of converting to a domain type. Or something.

If people understand "equivalent", then just ignore me.

**Review by greg-szabo — APPROVED** [review]

**Pull request merged**